### ISSUE TITLE
Support partial beta diversity calculations

### ISSUE THREAD

**Issue by wasade** [biocore/scikit-bio#1221]
Currently, the beta diversity of all pairwise combinations of sample IDs are computed. However, it would be nice to compute only a partial set in the case of combining studies (i.e., only compute the necessary distances to merge a set of samples) or if performing a block-oriented computation of the subsequent full distance matrix to avoid duplicated compute. 

A possible modification to the `beta_diversity` API would be to allow `ids` to either be a iterable of `str` or an iterable of `tuple`. In the former case, all pairwise combinations would be computed. In the later case, only the distances between the described pairs would be computed. It is not clear though if and how this should impact the return.

**Comment by josenavas** [comment]
Just want to ping on this issue as I think the potential use cases for this are increasing.
@gregcaporaso @jairideout @wasade @ebolyen which is the best way of starting a formal discussion to get this API changes into scikit-bio?

**Comment by wasade** [comment]
Thanks, @josenavas, for following up on this. 

From looking at the code, I think I may have a solution that does not necessitate an API change:

```python
def partial_pw(id_pairs, counts, metric, **kwargs):
    n_ids = len({a for a, b in id_pairs}.union({b for a, b in id_pairs}))
    dm = np.zeros((n_ids, n_ids), dtype=float)
    # do some fancy indexing that isn't shown here
    for u, v in id_pairs_indexed:
        dm[u, v] = metric(counts[u], counts[v], **kwargs)
    return dm

# ...then as a user/developer
pw = functools.partial(partial_pw, my_id_pairs)
ids = {a for a, b in my_id_pairs}.union({b for a, b in my_id_pairs})
partial_dm = beta_diversity('unweighted_unifrac', counts, ids=ids, validate=True, pairwise_func=pw, tree=tree, otu_ids=otu_ids)
```

I think this would accomplish what is necessary. So then the only addition to the public API would be a method `partial_pw` which is metric agnostic. There would not need to be any modification to existing APIs. 

If this seems reasonable, I can issue a PR on this. I'm not too jazzed (yet) on the ID ordering, as it would be important to minimize any juggling when merging DMs in the case of a block decomposition.

**Comment by gregcaporaso** [comment]
@wasade, I think that would work. I guess the main question is would we want to support this through the ``pairwise_func``, or though the ``beta_diversity`` interface directly. Note that we can add new parameters to ``beta_diversity``, so it seems like we might be able to build this into the existing API and take some burden off the user.

**Comment by wasade** [comment]
Excellent point. So perhaps the following?

```python
def beta_diversity(..., id_pairs=None, **kwargs):
    """
    ...
    
    Parameters
    ----------
    ...
    id_pairs : iterable of tuple, optional
        An iterable of tuples of IDs to compare (e.g., [('a', 'b'), ('a', 'c'), ...]). If specified, the set of IDs described must be a subset of `ids`. `id_pairs` is a mutually exclusive argument with `pairwise_func`.
    ...
    """
    ...
    if id_pairs is not None:
        # verify ids is set and is a subset of ids
        # verify pairwise_func is not set
        # construct pw method
    ...
```

**Comment by wasade** [comment]
Just a note, working on a PR right now so we can see what this looks like.

**Comment by wasade** [comment]
@gregcaporaso, ran into a wall: supporting metrics as `str` is not trivial as it doesn't looks like scipy exposes their [lookup](https://github.com/scipy/scipy/blob/v0.17.0/scipy/spatial/distance.py#L1249). UniFarc and `callable` should be fine, however. Curious to hear your thoughts on this. Currently circumventing this by not supporting `str` except for unifrac as it's implicit

**Comment by wasade** [comment]
HAH! The 1337 hax0r PR ftw

**Comment by ebolyen** [comment]
Out of curiosity how is this different from something like 

`map(metric, map(lambda a, b: (counts[a], counts[b]), id_pairs))`? 

Are the ID pairs always "complete" in that it produces a subset of the all-pairwise distance matrix?

If so, wouldn't the ideal solution be to allow view-based subsetting of the eventual  `FeatureTable`?

**Comment by wasade** [comment]
I don't think it is different. 

The expectation, though not stated, is that the resulting distance matrix is a subset of the all-pairwise one such that `reduce(add, partial_dms) == full_dm` assuming all pairs were enumerated.

A view is ideal as well, but not sufficient by itself. In the case of the block decomposition method (POC [here](https://github.com/wasade/yummy-octo-duck/blob/master/ipynb/block%20distance%20matrix%20compute.ipynb)), in the absence of partial support, you end up recomputing elements of the resulting matrix a large number of times. I don't recall how it scaled but it was a huge waste.

Just as a note, in the case of UniFrac, you want both a view as well as the subset of the tree.

**Comment by ebolyen** [comment]
Thanks for the explanation @wasade, makes sense why subsetting wouldn't necessarily work in the optimized case.

Given that, what is it that a partial beta diversity returns? It is kind of a "jagged"/"incomplete" distance matrix. This would also prevent us from just adding a parameter to `beta_diversity` as the return type is dependant unless we changed how `DistanceMatrix` works.

**Comment by wasade** [comment]
Right now, it is returning a full distance matrix that is zero'd for all non computed pairs. That is not ideal, but easiest to drop in. For [instance](https://github.com/biocore/scikit-bio/pull/1337/files#diff-220a2f5681654199aa60a12904851b0bR618)

I like the functional approach particularly as it circumvents the construction of what is likely a sparse matrix (which would be bad in the block decomposition method). What's missing is how to go from the resulting condensed form back to the matrix, which is I think what you're getting at.

**Comment by wasade** [comment]
Sorry! It's the full distance matrix then run through `squareform` to return the condensed form, so that it can then be fed into `DistanceMatrix`...

**Comment by ebolyen** [comment]
Right, it seems that something like an alternate constructor on `DistanceMatrix` "might" be a better way to recompose the fragments. OR we allow `DistanceMatrix` to exist as a fragment, which you could just add together, like your `reduce` example.

**Comment by ebolyen** [comment]
Although reducing would probably be just as inefficient as dealing with zeros.

**Comment by wasade** [comment]
I'm intrigued by the condensed form though as it should be possible to do:

```python
condensed_partials = voodoo
DistanceMatrix(np.sum.reduce(condensed_partials), ids)
```

...Still have a large number of zeros, but at least it's `((m * m) - m) / 2)`...

**Comment by wasade** [comment]
The method I did in the POC did full DMs but they were only of size `len(ids)`. The full matrix was then reassembled from the individual blocks.

The reduction wouldn't actually work anyway as `ids` needs to correspond to `counts`, yet the reduction assumes that `ids` represents the full set and not just those being computed on.

**Comment by wasade** [comment]
I believe #1337 is good for close review. I have one open question about example documentation. 

Separate from this specific issue, but just to make a note for those reading along, once #1337 is merged, I'll issue a PR for the block decomposition method for beta diversity. I'll also issue a separate PR with an optimized derivative of `TreeNode.shear`, which is a critical performance optimization for massive trees, UniFrac and the block decomposition method.

**Comment by jairideout** [comment]
Fixed in #1337.
